README.md
# simple_arm [![Build Status](http://build.ros.org/buildStatus/icon?job=Kbin_uX64__simple_arm__ubuntu_xenial_amd64__binary)](http://build.ros.org/job/Kbin_uX64__simple_arm__ubuntu_xenial_amd64__binary)

Simple velocity controlled arm. Teleoperation software and firmware.

![simple_arm](http://wiki.ros.org/simple_arm?action=AttachFile&do=get&target=simple+arm+joystick+diagram.png)

## Installation

```
$ sudo apt-get install ros-kinetic-simple-arm
```

**Full documentation on wiki: [http://wiki.ros.org/simple_arm](http://wiki.ros.org/simple_arm)**


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package simple_arm
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.1.0 (2017-07-20)
------------------
* First release

arm_firmware/src/main.cpp
#include <Arduino.h>
#include <Servo.h>

// PWM specs of the Victor SP motor controller.
//      https://www.vexrobotics.com/217-9090.html
#define victorMax 2350
#define victorMin 650

struct JOINTPINS{
  int wrist_roll = 9; // wrist roll pin
  int wrist_pitch = 10; // wrist pitch pin
  int upper_elbow = 11; // upper elbow pin
  int lower_elbow = 12;  // lower elbow pin
  int base_yaw = 13;  // base yaw pin
  int grip_enable = A4;  // gripper enable pin
  int grip_open = A5;  // gripper open pin
  int grip_close = A6;  // gripper close pin
  int cam_tilt = 7; // camera servo pin
}pins;

struct RECIEVED{
  float wrist_roll,wrist_pitch,upper_elbow,lower_elbow,base_yaw,gripper,winch,cam_tilt;
}data;

Servo joint0, joint1, joint2, joint3, joint4, cam_tilt;

int mapToVictor(float input){
  return map((input*100),-100,100,victorMin,victorMax);
}
void recievedData(){
  if(Serial.available()>=sizeof(float)){
    Serial.readBytes((char*)&data.gripper,sizeof(float));
    Serial.readBytes((char*)&data.wrist_roll,sizeof(float));
    Serial.readBytes((char*)&data.wrist_pitch,sizeof(float));
    Serial.readBytes((char*)&data.upper_elbow,sizeof(float));
    Serial.readBytes((char*)&data.lower_elbow,sizeof(float));
    Serial.readBytes((char*)&data.base_yaw,sizeof(float));
    Serial.readBytes((char*)&data.cam_tilt,sizeof(float));
  }
}

void writeToJoints(){
  joint0.writeMicroseconds(mapToVictor(data.wrist_roll));
  joint1.writeMicroseconds(mapToVictor(data.wrist_pitch));
  joint2.writeMicroseconds(mapToVictor(data.upper_elbow));
  joint3.writeMicroseconds(mapToVictor(data.lower_elbow));
  joint4.writeMicroseconds(mapToVictor(data.base_yaw));
  cam_tilt.write((int)data.cam_tilt);

  // Gripper Open
  if (data.gripper == 1){
    digitalWrite(pins.grip_enable, HIGH);   // enable pin on
    digitalWrite(pins.grip_open, HIGH);   // turn open on
    digitalWrite(pins.grip_close, LOW);   // turn close off
  }
  // Gripper Close
  else if (data.gripper == -1){
    digitalWrite(pins.grip_enable, HIGH);   // enable pin on
    digitalWrite(pins.grip_open, LOW);   // turn open off
    digitalWrite(pins.grip_close, HIGH);   // turn close on
  }
  // Gripper Stop
  else if (data.gripper == 0){
    digitalWrite(pins.grip_enable, LOW);   // enable pin off
    digitalWrite(pins.grip_open, LOW);   // turn open off
    digitalWrite(pins.grip_close, LOW);   // turn close on
  }
}
void setup() {
  Serial.begin(9600);
  joint0.attach(pins.wrist_roll);
  joint1.attach(pins.wrist_pitch);
  joint2.attach(pins.upper_elbow);
  joint3.attach(pins.lower_elbow);
  joint4.attach(pins.base_yaw);
  cam_tilt.attach(pins.cam_tilt);
  pinMode(pins.grip_enable, OUTPUT);
  pinMode(pins.grip_open, OUTPUT);
  pinMode(pins.grip_close, OUTPUT);
  recievedData();
}

void loop() {
  recievedData();
  writeToJoints();
}

src/simple_arm/simple_arm_serial.py
#!/usr/bin/python

import rospy
import sensor_msgs.msg
import serial
import struct

RECOGNIZED_ARM = ['hip', 'shoulder', 'elbow', 'lower_elbow']
RECOGNIZED_GRIPPER = ['wrist', 'grip']
ALL_RECOGNIZED_JOINTS = []
ALL_RECOGNIZED_JOINTS.extend(RECOGNIZED_ARM)
ALL_RECOGNIZED_JOINTS.extend(RECOGNIZED_GRIPPER)


def on_new_joint(data):
  # For better readability and robustness, convert ROS joint message
  # to dict so that joint values can be access by name instead of by index.
  joint_position = {}
  for idx, name in enumerate(data.name):
    joint_position[name] = data.position[idx]

  # Execute arm position
  if data.name[0:4] == RECOGNIZED_ARM:
    encoded_position = struct.pack("<Bffff", 1,
                                    joint_position['shoulder'],
                                    joint_position['elbow'],
                                    joint_position['lower_elbow'],
                                    joint_position['hip'])
    theSerial.write(encoded_position)
    rospy.loginfo('Executed arm position %s'% data)

  # Execute gripper position
  if data.name[4:6] == RECOGNIZED_GRIPPER:
    encoded_position = struct.pack("<Bff", 3,
                                     joint_position['wrist'],
                                     joint_position['grip'])
    theSerial.write(encoded_position)
    rospy.loginfo('Executed gripper position %s'% data)

  unknown_joints = [joint for joint in data.name if joint not in ALL_RECOGNIZED_JOINTS]
  if unknown_joints:
    rospy.logwarn('Ignoring unknown joints: %s' % unknown_joints)

theSerial = serial.Serial()
theSerial.port = "/dev/ttyACM0" # SET SERIAL DEVICE
theSerial.open()

rospy.init_node("joint_sender")
subscriber = rospy.Subscriber("/joint_states", sensor_msgs.msg.JointState, on_new_joint, queue_size=10)
rospy.spin()




src/simple_arm/simple_arm_teleop.py
#!/usr/bin/python
import rospy
import serial
import struct
import numpy as np

from sensor_msgs.msg import Joy

class JoyArmSerial:
    def __init__(self):
        self.velocities = {
            'grip': 0,
            'wrist_roll': 0,
            'wrist_pitch': 0,
            'upper_elbow': 0,
            'lower_elbow': 0,
            'base_yaw': 0
        }
        self.positions = {
            'camera': 0
        }
        baudrate = rospy.get_param('~baudrate', 9600)
        self.serialDev = serial.Serial(baudrate=baudrate)
        self.serialDev.port = rospy.get_param("~serial_dev")
        self.serialDev.open()
        self.arm_sub = rospy.Subscriber("/joy_arm", Joy, controller.arm_joy_callback)

    def write_serial(self):
      # Execute arm position
        rospy.loginfo('velocities:%s\n' % self.velocities)
        rospy.loginfo('positions:%s\n' % self.positions)
        encoded_position = struct.pack("<fffffffff",
                                        self.velocities['grip'],
                                        self.velocities['wrist_roll'],
                                        self.velocities['wrist_pitch'],
                                        self.velocities['upper_elbow'],
                                        self.velocities['lower_elbow'],
                                        self.velocities['base_yaw'],
                                        self.positions['camera']
                                        )
        self.serialDev.write(encoded_position)

    def arm_joy_callback(self, data):
        self.velocities['grip'] = 0
        self.velocities['wrist_roll'] = 0
        self.velocities['wrist_pitch'] = 0
        self.velocities['upper_elbow'] = 0
        self.velocities['lower_elbow'] = 0
        self.velocities['base_yaw'] = 0

        # button 12 = double speed
        SPEED_MULTIPLIER = 1
        if data.buttons[11]:
            SPEED_MULTIPLIER = 2
        # button 10 = half speed
        if data.buttons[9]:
            SPEED_MULTIPLIER = 0.5

        # +big stick twist = base clockwise
        # -big stick twist = base counter-clockwise
        self.velocities['base_yaw'] = data.axes[2] / 5.0 * SPEED_MULTIPLIER

        # +big stick forward = big arm up
        # -big stick back = big arm down
        self.velocities['lower_elbow'] = data.axes[1] / 3.0 * SPEED_MULTIPLIER

        # thumb stick forward = small arm down
        # thumb stick back = small arm up
        self.velocities['upper_elbow'] = data.axes[5] / 5.0 * SPEED_MULTIPLIER

        # paddle forward = arm camera up
        # paddle back = arm camera down
        self.positions['camera'] = np.interp(data.axes[3],[-1,1],[7,77]) # this value is angular postion in degrees

        # button 9 = wrist clockwise
        # button 11 = wrist counter-clockwise
        if data.buttons[8]:
            self.velocities['wrist_roll'] = 0.35 * SPEED_MULTIPLIER
        if data.buttons[10]:
            self.velocities['wrist_roll'] = -0.35 * SPEED_MULTIPLIER

        # button 6 = wrist down
        # button 4 = wrist up
        if data.buttons[4]:
            self.velocities['wrist_pitch'] = 0.15 * SPEED_MULTIPLIER
        if data.buttons[2]:
            self.velocities['wrist_pitch'] = -0.15 * SPEED_MULTIPLIER

        # trigger = close gripper
        # thumb button = open gripper
        if data.buttons[0]:
            self.velocities['grip'] = 1
        if data.buttons[1]:
            self.velocities['grip'] = -1

        # MOVE ARM
        self.write_serial()


def main():
    rospy.init_node("joystick_teleoperation_arm")
    controller = JoyArmSerial()
    rospy.spin()


src/simple_arm/simple_arm.py
#!/usr/bin/python
import rospy
import serial
import struct
import numpy as np

from sensor_msgs.msg import Joy

class SimpleArm:
    def __init__(self):
        self.velocities = {
            'grip': 0,
            'wrist_roll': 0,
            'wrist_pitch': 0,
            'upper_elbow': 0,
            'lower_elbow': 0,
            'base_yaw': 0
        }
        self.positions = {
            'camera': 0
        }
        baudrate = rospy.get_param('~baudrate', 9600)
        self.serialDev = serial.Serial(baudrate=baudrate)
        self.serialDev.port = rospy.get_param("~microcontroller_serial_device")
        self.serialDev.open()
        self.arm_sub = rospy.Subscriber("/joy_arm", Joy, self.arm_joy_callback)

    def write_serial(self):
      # Execute arm position
        rospy.loginfo('velocities:%s\n' % self.velocities)
        rospy.loginfo('positions:%s\n' % self.positions)
        encoded_position = struct.pack("<fffffff",
                                        self.velocities['grip'],
                                        self.velocities['wrist_roll'],
                                        self.velocities['wrist_pitch'],
                                        self.velocities['upper_elbow'],
                                        self.velocities['lower_elbow'],
                                        self.velocities['base_yaw'],
                                        self.positions['camera']
                                        )
        self.serialDev.write(encoded_position)

    def arm_joy_callback(self, data):
        self.velocities['grip'] = 0
        self.velocities['wrist_roll'] = 0
        self.velocities['wrist_pitch'] = 0
        self.velocities['upper_elbow'] = 0
        self.velocities['lower_elbow'] = 0
        self.velocities['base_yaw'] = 0

        # button 12 = hold button to double motor speed
        SPEED_MULTIPLIER = 1
        if data.buttons[11]:
            SPEED_MULTIPLIER = 2
        # button 10 = hold button to half motor speed
        if data.buttons[9]:
            SPEED_MULTIPLIER = 0.5

        # big stick twist = base clockwise
        # big stick twist = base counter-clockwise
        self.velocities['base_yaw'] = data.axes[2] / 5.0 * SPEED_MULTIPLIER

        # big stick forward = lower elbow up
        # big stick back = lower elbow down
        self.velocities['lower_elbow'] = data.axes[1] / 3.0 * SPEED_MULTIPLIER

        # thumb stick forward = upper elbow down
        # thumb stick back = upper elbow up
        self.velocities['upper_elbow'] = data.axes[5] / 5.0 * SPEED_MULTIPLIER

        # paddle forward = arm camera up
        # paddle back = arm camera down
        self.positions['camera'] = np.interp(data.axes[3],[-1,1],[7,77]) # this value is angular postion in degrees

        # button 9 = wrist clockwise
        # button 11 = wrist counter-clockwise
        if data.buttons[8]:
            self.velocities['wrist_roll'] = 0.35 * SPEED_MULTIPLIER
        if data.buttons[10]:
            self.velocities['wrist_roll'] = -0.35 * SPEED_MULTIPLIER

        # button 6 = wrist down
        # button 4 = wrist up
        if data.buttons[4]:
            self.velocities['wrist_pitch'] = 0.15 * SPEED_MULTIPLIER
        if data.buttons[2]:
            self.velocities['wrist_pitch'] = -0.15 * SPEED_MULTIPLIER

        # trigger = close gripper
        # thumb button = open gripper
        if data.buttons[0]:
            self.velocities['grip'] = 1
        if data.buttons[1]:
            self.velocities['grip'] = -1

        # MOVE ARM
        self.write_serial()


def main():
    rospy.init_node("simple_arm")
    controller = SimpleArm()
    rospy.spin()
